frontend/src/metabase/dashboard/dashboard.js

```diff
@@ -1173,6 +1162,28 @@ const loadingDashCards = handleActions(
   { dashcardIds: [], loadingIds: [], startTime: null },
 );
 
+// This loads database metadata and nested questions for a list of dash cards.
+// We load the nested questions separately because the db metadata only includes real tables.
+function fetchDashCardEntities(dispatch, dashCards) {
```

Comment:
Two things:

1. I'm not sure but it seems like `fetchDatabaseMetadata` might be overfetching tables that aren't needed. Maybe we should be calling `Tables.actions.fetchMetadata` for non-nested queries too.
2. We have an action called `loadMetadataForQuery` we use in the query builder: https://github.com/metabase/metabase/blob/03a5f98ff2c449ac30e02b6762cb0a37f2ff4449/frontend/src/metabase/redux/metadata.js#L326 I don't think we'll want to use that directly since we want to deduplicate requests, but we might want to write a similar one called `loadMetadataForQueries` or something, that calls `dependentMetadata()` on each query, deduplicates them, and makes the `Tables.actions.fetchMetadata` requests.

Reply:
Good call! That feels like a better solution.

I'm somewhat nervous that replacing the one big db fetch with one per table could slow things down for some dashboards. Still it's a step in the right direction though. If we see any slowness, we could batch those table fetches.